src/core/algorithms/fd/dynfd/model/dynamic_relation_data.cpp

```diff
@@ -0,0 +1,265 @@
+}
+
+std::unique_ptr<DynamicRelationData> DynamicRelationData::CreateFrom(
+        config::InputTable const& input_table) {
+    auto schema = std::make_unique<RelationalSchema>(input_table->GetRelationName());
+    int next_value_id = 1;
+    size_t const num_columns = input_table->GetNumberOfColumns();
+    ValueDictionary value_dictionary(num_columns);
+    std::vector<std::vector<int>> column_dictionary_encoded_data =
+            std::vector<std::vector<int>>(num_columns);
+    std::vector<std::vector<int>> compressed_records;
+
+    while (input_table->HasNextRow()) {
+        std::vector<std::string> row = input_table->GetNextRow();
+
+        if (row.size() != num_columns) {
+            LOG(WARNING) << "Unexpected number of columns for a row, skipping (expected "
+                         << num_columns << ", got " << row.size() << ")";
+            continue;
+        }
+
+        compressed_records.emplace_back(num_columns);
+
+        for (size_t index = 0; index < row.size(); ++index) {
+            std::string const& field = row[index];
+
+            auto location = value_dictionary[index].find(field);
+            int value_id;
+            if (location == value_dictionary[index].end()) {
+                value_id = next_value_id++;
+                value_dictionary[index][field] = -static_cast<int>(compressed_records.size());
+
+                compressed_records.back()[index] = -value_id - 1;
+            } else {
+                value_id = location->second;
+
+                if (value_id < 0) {
+                    int column_ind = -value_id - 1;
+                    value_id = -compressed_records[column_ind][index] - 1;
+                    location->second = value_id;
+                    compressed_records[column_ind][index] = value_id;
+                }
+
+                compressed_records.back()[index] = value_id;
+            }
+
+            column_dictionary_encoded_data[index].push_back(value_id);
+        }
+    }
+
+    std::vector<CompressedColumnData> column_data;
+    for (size_t i = 0; i < num_columns; ++i) {
+        auto column = Column(schema.get(), input_table->GetColumnName(i), i);
+        schema->AppendColumn(std::move(column));
+        auto pli = DynamicPositionListIndex::CreateFor(column_dictionary_encoded_data[i], i);
+        column_data.emplace_back(schema->GetColumn(i), std::move(pli));
+    }
+
+    size_t next_record_id = (!column_data.empty() ? column_data[0].GetNumRows() : 0);
+    std::vector<size_t> all_ids(next_record_id);
+    std::iota(all_ids.begin(), all_ids.end(), 0);
+
+    input_table->Reset();
+
+    return std::make_unique<DynamicRelationData>(std::move(schema), std::move(column_data),
+                                                 std::unordered_set(all_ids.begin(), all_ids.end()),
+                                                 std::move(value_dictionary), next_value_id,
+                                                 next_record_id, std::move(compressed_records));
+}
```

Comment:
On the left we have `CreateFrom`, on the right `InsertBatch`:
<img width="2174" height="855" alt="image" src="https://github.com/user-attachments/assets/32a704f6-589d-4098-a46f-87af21466fce" />
Code looks virutally the same inside. Same for the `InsertRecordsFromUpdateBatch`. Extract common logic into a helper method, please